jwst/ramp_fitting/ramp_fit.py

```diff
@@ -303,92 +328,237 @@ def ols_ramp_fit(model, buffsize, save_opt, readnoise_model, gain_model,
+                             inv_var_both4_int.shape[2]))
+
+        num_seg_int = num_seg_per_int[num_int,:, :] # number of segs per pixel
+        num_pix = num_seg_int.shape[0] * num_seg_int.shape[1]
+        num_seg_int1 = num_seg_int.reshape(num_pix) # [ seg, 1D pixel ]
+
+        arr_ind_all = np.array([np.arange( var_p4_int2.shape[0] ),] *
+                                num_pix).transpose()
+
+        var_p4_int2[ arr_ind_all >= num_seg_int1 ] = 0
+        var_r4_int2[ arr_ind_all >= num_seg_int1 ] = 0
+        inv_var_both4_int2[ arr_ind_all >= num_seg_int1 ] = 0
+
+        var_p4_int = var_p4_int2.reshape((var_p4_int.shape[0],
+                                          var_p4_int.shape[1],
+                                          var_p4_int.shape[2]))
+
+        var_r4_int = var_r4_int2.reshape((var_r4_int.shape[0],
+                                          var_r4_int.shape[1],
+                                          var_r4_int.shape[2]))
+
+        inv_var_both4_int = inv_var_both4_int2.reshape((inv_var_both4_int.shape[0],
+                        inv_var_both4_int.shape[1], inv_var_both4_int.shape[2]))
+
+        # Zero out non-existing segments
+        var_p4_int *= ( segs_4[num_int,:,:,:] > 0)
+        var_r4_int *= ( segs_4[num_int,:,:,:] > 0)
+        inv_var_both4_int *= ( segs_4[num_int,:,:,:] > 0)
+
+        s_inv_var_both3[num_int,:,:] = (inv_var_both4[num_int,:,:,:]).sum(axis=0)
+        var_both3[num_int, :, :] = 1./s_inv_var_both3[num_int, :, :]
+
+    var_p3 = utils.rem_nan_inf(var_p3)
+    var_p3[ var_p3 < 0. ] = 0.
+
+    var_r3 = utils.rem_nan_inf(var_r3)
+
+    var_both3 = utils.rem_nan_inf(var_both3)
+    var_both3[ var_both3 < 0.] = 0.
+
+    # Now that the segment-specific and integration-specific variances have
+    #   been calculated, the segment-specific, integration-specific, and
+    #   overall slopes will be calculated. The integration-specific slope is
+    #   calculated as a weighted average of the segments in the integration:
+    #     slope_int = sum_over_segs(slope_seg/var+seg)/ sum_over_segs(1/var_seg)
+    #  The overall slope is calculated as a weighted average of the segments in
+    #     all integrations:
+    #     slope = sum_over_integs_and_segs(slope_seg/var+seg)/
+    #                    sum_over_integs_and_segs(1/var_seg)
+
+    slope_seg = opt_res.slope_seg.copy()
+    slope_by_var4 = slope_seg / var_both4
+    slope_by_var4 = utils.rem_nan_inf(slope_by_var4)
+
+    s_slope_by_var3 = slope_by_var4.sum(axis=1) # sum over SEGMENTS (not integs)
+    s_slope_by_var2 = s_slope_by_var3.sum(axis=0) # sum over integrations
+
+    s_inv_var_both2 = s_inv_var_both3.sum(axis=0)
+
+    # Compute the 'dataset-averaged' slope
+    slope_dataset2 = s_slope_by_var2/s_inv_var_both2
+    slope_dataset2 = utils.rem_nan_inf(slope_dataset2)
+
+    # Compute the integration-specific slope
+    the_num = (slope_seg * inv_var_both4).sum(axis = 1) # sum over all segments
```

Comment:
I don't recall exactly what we decided during discussions a month or so ago, but I think it does make some sense to NOT include Poisson noise in the weighting, because that would have a tendency for a bias towards lower slope values (which have lower Poisson noise and hence higher weight).

Reply:
@hbushouse If the median slope over the entire exposure is used for computing the Poisson noise, that should eliminate the bias toward lower slopes, shouldn't it (because the same slope is used for every segment)?  I thought the bias that was of concern had to do with the lengths of the segments, though I don't understand why that would introduce a bias.

Reply:
Yes, I believe you're right and I think that's why and where we wanted to use the median slopes (from first differences) - it was for use in estimating the Poisson noise. As opposed to using the actual fitted slopes, which vary from segment to segment and integration to integration. So in that case it should be OK to include variances from both Poisson and read noise in the weighting for computing averaged slopes?

Reply:
I agree. It is my recollection of our earlier discussion that that was the reason that we agreed to use both Poisson and readnoise.